src/main/tut/2.4-representable-functors.md

```diff
@@ -0,0 +1,208 @@
+    def fmap[X, B](f: X => B)(h: Reader[A, X])
+        : Reader[A, B] =
+      Reader(x => f(h.run(x)))
+  }
+
+/* type lambda with Kind Projector plugin:
+   new Functor[Reader[A, ?]] {...} */
+```
+................
+```Haskell
+type Op a x = x -> a
+```
+```tut:silent
+case class Op[A, X](f: X => A)
+```
+................
+```Haskell
+instance Contravariant (Op a) where
+    contramap f h = h . f
+```
+```scala
+implicit def opContravariant[A] =
+  new Contravariant[({type T[X] = Op[A, X]})#T] {
+    def contramap[X, B](f: B => X)
+        : Op[A, X] => Op[A, B] = {
+      case Op(h) => Op(h compose f)
+    }
+  }
+
+/* type lambda with Kind Projector plugin:
+   new Contravariant[Op[A, ?]] {...} */
+```
+................
+```Haskell
+instance Profunctor (->) where
+  dimap ab cd bc = cd . bc . ab
+  lmap = flip (.)
+  rmap = (.)
+```
+```scala
+implicit def arrowProfunctor =
+  new Profunctor[Function1] {
+    def dimap[A, B, C, D]
+        (ab: A => B)(cd: C => D)(bc: B => C)
+        : A => D =
+      cd compose bc compose ab
+
+    def lmap[A, B, C]
+        (f: C => A)(ab: A => B): C => B =
+      f andThen ab
+
+    def rmap[A, B, C]
+        (f: B => C)(ab: A => B): A => C =
+      f compose ab
+  }
+```
+................
+```Haskell
+C(-, =) :: Cop × C -> Set
+```
+```scala
+val `C(-, =)`: (Cop `×` C) => Set
+```
+................
+```Haskell
+αx :: C(a, x) -> F x
+```
+```scala
+val αx: `C(A, X)` => F[X]
+```
+................
+```Haskell
+alpha :: forall x. (a -> x) -> F x
+```
+```scala
+def alpha[X]: (A => X) => F[X]
+```
+................
+```Haskell
+fmap f . alpha = alpha . fmap f
+```
+```scala
+(fmap(f) _ compose alpha) ==
+  (alpha compose fmap(f))
+```
+................
+```Haskell
+fmap f (alpha h) = alpha (f . h)
+```
+```scala
+fmap(f)(alpha(h)) == alpha(f compose h)
+```
+................
+```Haskell
+beta :: forall x. F x -> (a -> x)
+```
+```scala
+def beta[X]: F[X] => (A => X)
+```
+................
+```Haskell
+alpha :: forall x. (Int -> x) -> [x]
+alpha h = map h [12]
+```
+```tut:silent
+def alpha[X]: (Int => X) => List[X] =
+  h => List(12).map(h)
+```
+................
+```Haskell
+map f (map h [12]) = map (f . h) [12]
+```
+```scala
+fmap(f)(fmap(h)(List(12))) ==
+  fmap(f compose h)(List(12))
+
+/* Or using scala stdlib:
+List(12).map(h).map(f) ==
+  List(12).map(f compose h) */
+```
+................
+```Haskell
+beta :: forall x. [x] -> (Int -> x)
+```
+```scala
+def beta[X]: List[X] => (Int => X)
+```
+................
+```Haskell
+class Representable f where
+   type Rep f :: *
+   tabulate :: (Rep f -> x) -> f x
+   index    :: f x -> Rep f -> x
+```
+```tut:silent
+trait Representable[F[_]] {
+  type Rep[F[_]]
+
+  def tabulate[X](f: Rep[F] => X): F[X]
+
+  def index[X]: F[X] => Rep[F] => X
+}
+```
+................
+```Haskell
+data Stream x = Cons x (Stream x)
+```
+```tut:silent
+sealed trait Stream[A]
+final case class Cons[A](
+    h: () => A,
+    t: () => Stream[A]
+    ) extends Stream[A]
+```
```

Comment:
This is a general thing about datatypes, but in scala one would never have a trait with only one subclass, just go with 
```scala
case class Stream[X](h: X, rest: Stream[X])
```